smart-contracts/contracts/Airdrops.sol

```diff
@@ -0,0 +1,187 @@
+ * The contract allows only those who have signed the campaign's Terms of Service (TOS) to claim tokens.
+ * The owner sets up campaigns with a TOS and a corresponding Merkle root representing eligible entries.
+ */
+contract Airdrops is IAirdrops {
+  /// @notice Structure representing a campaign's Terms of Service and its corresponding Merkle tree root.
+  struct Campaign {
+    string tos;
+    bytes32 merkleRoot;
+  }
+
+  /// @notice The ERC20 token to be airdropped (UP)
+  IERC20 public token;
+
+  /// @notice Owner for administrative functions
+  address public owner;
+
+  /// @notice Mapping from campaign name to campaign details.
+  mapping(string => Campaign) public campaigns;
+
+  /// @notice Mapping from campaign name to addresses that have signed the TOS.
+  mapping(string => mapping(address => bool)) public signedTos;
+
+  /// @notice Mapping of Merkle tree leaves that have been claimed to prevent double claiming.
+  mapping(bytes32 => bool) public claimedLeafs;
+
+  /// @notice Emitted when a campaign is set.
+  event CampaignSet(string indexed campaignName, bytes32 merkleRoot);
+  /// @notice Emitted when a recipient signs the TOS.
+  event TosSigned(string indexed campaignName, address indexed recipient);
+  /// @notice Emitted when tokens are claimed.
+  event TokensClaimed(string indexed campaignName, address indexed recipient, uint256 amount);
+
+  /**
+   * @notice Constructor that sets the token to be airdropped and initializes the owner.
+   * @param _token The address of the ERC20 token contract (UP).
+   */
+  constructor(address _token) {
+    token = IERC20(_token);
+    owner = msg.sender;
+  }
+
+  /**
+   * @notice Modifier to restrict functions to the contract owner.
+   */
+  modifier onlyOwner() {
+    if (msg.sender != owner) revert OnlyOwner();
+    _;
+  }
+
+  /**
+   * @notice Sets the Merkle root and Terms of Service (TOS) for a campaign.
+   * @param campaignName The name/identifier of the campaign.
+   * @param tos The Terms of Service text for the campaign.
+   * @param root The Merkle tree root corresponding to eligible entries.
+   */
+  function setMerkleRootForCampaign(
+    string calldata campaignName,
+    string calldata tos,
+    bytes32 root
+  ) external override onlyOwner {
+    campaigns[campaignName] = Campaign(tos, root);
+    emit CampaignSet(campaignName, root);
+  }
+
+  /**
+   * @notice Signs the Terms of Service for a given campaign.
+   *
+   * The signature must be produced off-chain, signing the Ethereum Signed Message hash of the campaign's TOS.
+   *
+   * @param campaignName The campaign identifier.
+   * @param recipient The address that is signing the TOS.
+   * @param signature The signature bytes (generated off-chain) of the hash of the campaign's TOS.
+   */
+  function signTos(
+    string calldata campaignName,
+    address recipient,
+    bytes calldata signature
+  ) external override {
+    if (msg.sender != recipient) revert NOT_AUTHORIZED();
```

Comment:
I think it's ok to have a different recipient and msg.sender as long as the signature matches the recipient